### ISSUE TITLE
VPAT Reader: keyboard alternatives

### ISSUE THREAD

**Issue by abaevbog** [zotero/zotero#4224]
vpat 76 and 77: Annotation Tools  Lack Keyboard Alternatives

From the VPAT review: "In the Zotero desktop app's PDF Viewer, some of the annotation tools, such as the highlight and underline tools, do not have comprehensive keyboard alternatives. While keyboard alternatives may not be necessary for all annotation tools, commonly used tools should be accessible without relying on a pointing device. When these types of movements are essential to the function's design (e.g., the Draw Tool) and cannot have an accessible alternative in any way, they are exempt from the requirement of needing to be keyboard operable."

**Comment by abaevbog** [comment]
@mrtcode I am moving onto this issue, now that most easier reader-related vpat comments have been addressed, and I wanted to check in with you on the best way to approach this. Essentially, we need to have a way to manipulate annotations with a keyboard. My understanding is that the drawing tool is exempt from this but all remaining annotation types need to have an alternative.

My thinking is that "note", "text" and "area " annotations could be handled in similar ways: on keyboard-initiated click on the annotation button (or, perhaps with a keyboard shortcut?), we'll actually make and insert the annotation into a pre-defined location of the reader (e.g. top-right corner). From where one should be able to move the annotation when it has focus with arrowKeys (perhaps, with a modifier key, since `arrowUp`/`arrowDown` change annotation selection in the side bar. The main issue here is that I didn't find an existing way to move "text" and "area" annotations like we can move "note", so I suppose needs to be added as well. 

"highlight" and "underline" are a bit trickier since we need to know exactly where to place it in text. My idea is to assume that the user has some kind of screen reader (e.g. JAWS/Voiceover/NVDA) that can be used to select arbitrary text in browser mode. Then, I expect we should be able to know when text is selected from `selectionchange` event and apply the dark blue styling to indicate text is selected. From there, maybe a keyboard shortcut to turn selected text into the annotation?

These are my initial thoughts on how to approach this - I'll get started on piecing it together but let me know how this direction sounds!

**Comment by mrtcode** [comment]
This will require core changes in the PDF and DOM (EPUB, snapshot) views from me and @AbeJellinek. We need to figure out and synchronize approaches that would work in both view types.

**Creating annotations**

For creating highlight/underline annotations and selecting text, I suggest implementing a method similar to what Acrobat Reader does. This would involve enabling caret movement and text selection with the arrow keys, functioning like in a word processor. Just need to figure out how to start caret browsing mode with keyboard alone.

Acrobat Reader video where I firstly focus caret in the text with mouse and then move it with arrow keys:

https://github.com/user-attachments/assets/458e69f9-9db1-4fda-a009-460b72fd1e23

> My thinking is that "note", "text" and "area " annotations could be handled in similar ways: on keyboard-initiated click on the annotation button (or, perhaps with a keyboard shortcut?), we'll actually make and insert the annotation into a pre-defined location of the reader (e.g. top-right corner)

Yes, using a pre-defined location sounds good.

**Manipulating annotations**

For existing annotation manipulation I suggest using keyboard shortcuts (I haven't properly checked if they are interfering with something):

Resize highlight/underline annotations:
- <kbd>Cmd-ArrowLeft</kbd> / <kbd>Cmd-ArrowRight</kbd> to control the right resizing handle.
- <kbd>Cmd-Shift-ArrowLeft</kbd> / <kbd>Cmd-Shift-ArrowRight</kbd> to control the left resizing handle.

Note: I was considering supporting a single highlight/underline with multiple text ranges, but I’m not sure if this will be implemented as it might introduce too much unnecessary complexity (would need more than two resizing handles). Although, we are partially already doing this when a highlight spans into another page or when page headers/footers are excluded.

Move note, text, image and ink annotations:
- <kbd>Cmd-ArrowLeft</kbd>,  <kbd>Cmd-ArrowRight</kbd>,  <kbd>Cmd-ArrowUp</kbd>,  <kbd>Cmd-ArrowDown</kbd>

Resize text, image and ink annotations:
- <kbd>Cmd-ArrowLeft</kbd> / <kbd>Cmd-ArrowRight</kbd> to resize the right side.
- <kbd>Cmd-ArrowUp</kbd> / <kbd>Cmd-ArrowDown</kbd> to resize the bottom side (either won't be used for text annotation, or used to increase font size as well).
- Resizing left and top sides might not be necessary since the annotation can just be moved.

Change text annotation font size:
- <kbd>Cmd-Shift-Plus</kbd> reduce by 1pt.
- <kbd>Cmd-Shift-Minus</kbd> increase by 1pt.


> I didn't find an existing way to move "text" and "area" annotations like we can move "note", so I suppose needs to be added as well

Area annotation doesn’t have this because we decided that resizing to all four sides is sufficient, and we left the drag-and-drop functionality only for dropping the annotation into the note editor. Text annotation can be dragged in the same way as note annotation, except if the text is already focused, in which case you can only grab the dashed border.

**Comment by abaevbog** [comment]
Re text selection caret: that does make sense, I think it would be a useful feature in itself, not just from accessibility standpoint. I am not exactly sure how the caret would play with screen readers, to be honest, because I'm not familiar with the actual rendering part of the content enough and JAWS was causing issues in way simpler components. I hope something like `contenteditable` would work, and if not - we already have `aria-live` component that we can use to announce arbitrary announcements (e.g. which text was selected).

It does introduce a tricky question for screen readers users. Focus mode of screen readers would still be the best way to navigate the pages of PDF or snapshots (meaning, when screen reader is just "looking" at the content of the page without actually focusing anything and most keypresses are taken over by screen reader - e.g. arrowDown reads the next line instead of scrolling, arrowRight/left read the next word, etc.). It allows one to use a lot of screen-reader specific commands to navigate the content better (e.g. "Read the entire next paragraph" or "Go to the next heading" and such). In that case, the scenario I am thinking about is this: one navigates the reader with a virtual cursor in focus mode, runs into a sentence they want to annotate, and the question is, how could we place the cursor/annotation at that location?

Having played around with it, I can't so far find a reliable way to do it, since virtual cursor is internal to screen reader and we don't have access to it ... So I suppose we should just place it at the most reasonable location, such as the top of the current page, probably on tab from the document itself. Next tab could focus and navigate annotations.

**Comment by abaevbog** [comment]
> Could we instead make it so that Option-arrow keys "selects" the resizers of the annotation, and then arrow keys move the selected resizer?

Personally, I like this idea. Another candidate for "selecting" the way in which the annotation is being interacted with is a menu option. For instance, Tools > Move Annotation (only available when the annotation is focused) could make it so that arrows actually move the focused annotation, instead of focusing the next one? Similarly, Tools > Resize Annotation > Top/Right/Left/Bottom could select the mode where arrows move the border of the area. 

The only benefit of such approach is that it would have less "magic" keypresses one has to remember to perform an operation.

**Comment by abaevbog** [comment]
> Yes, using a pre-defined location sounds good.

Cool! Would dropping a note or text annotation on keyboard click of the annotation button into a predefined location of the reader be a good starting point for me? I don't suppose it would conflict with anything else, since the caret would not be involved?

**Comment by abaevbog** [comment]
That makes perfect sense! I got somewhat familiar with the reader's sidebar during earlier accessibility updates but `src/{pdf,dom}` are very much a mystery for me. So this outline makes perfect sense - I'll focus on the `common/` logic (e.g. `toolbar.js` for the keydown listener) that will just call the methods on the view to do the heavy lifting.

**Comment by mrtcode** [comment]
I think we shouldn’t rush the implementation yet, as there are other long-standing keyboard navigation conceptual gaps in the reader.

One important missing feature in the reader is the ability to quickly select any currently visible object (annotation, link, or in-text citation) in the view. If I remember correctly, this was functional in DOM views at some point and still works when running the dev version of the reader in a browser. Actually, it still scrolls into view and selects the last link in EPUB/snapshot if I press Shift-Tab from the context pane, which is also unexpected behavior.

In the PDF view, the original plan was to press TAB to activate a special selection mode (with a gray outline) and cycle through all currently visible elements in the view (annotations, links) using the arrow keys. This mode currently works, but only for annotations.

However, I think having two selection modes is too complicated.

Currently, regardless of whether you click in the view or the annotation sidebar, you still get the same annotation navigation with arrow keys. Instead, I’m thinking about separating these. We should keep the current behavior only when an annotation is focused in the sidebar, where you can navigate the 1D annotation list. But if you select an annotation or focus on a link in the view, you should navigate between the currently visible objects, possibly navigating in 2D space with all arrow keys.

To put it simply, if you want to navigate annotations, you select an annotation in the sidebar. But if you focus/select an object in the view, you navigate over all currently visible objects.

Also, as we introduce extra features, links and in-text citations should become as important to access as annotations.

Once an object is selected, pressing Enter triggers the default action:
- For annotations, it selects the annotation (and focuses on the comment if it’s empty).
- For external links, it opens the link.
- For internal links, and in-text citations, which normally have a popup on hover and navigate to the destination on click, it’s unclear which action to trigger as both are useful.

And then there should be an easy way to return back to object navigation, possibly by just pressing Escape.

**Comment by abaevbog** [comment]
@mrtcode gotcha! So, the idea is to essentially implement better keyboard navigation of annotations (and other objects) via keyboard before jumping straight to adding/manipulating them without using a mouse? It makes sense. I did notice two different annotation navigation modes when looking at earlier accessibility issues with the grey outline navigation mode kicking mainly when the sidebar is collapsed. 

Am I correctly understanding that most of this keyboard navigation logic is implemented in `src/{pdf,dom}` views? I totally understand that views-specific changes are easier to coordinate for you with @AbeJellinek. But let me know if there is any specific aspect here that I could help with to get to resolving this vpat issue sooner! Otherwise, we can revisit this particular accessibility question later, once the keyboard navigation in the reader is improved.

**Comment by mrtcode** [comment]
If it is a priority to support creating/modifying annotations, we can try to implement that first. However, we need to decide exactly how that will work, either by 1) using keyboard shortcuts to resize annotations, 2) selecting specific resizers, as suggested by @AbeJellinek, or 3) considering other methods. It would be good to investigate what other software is doing.

If we only had square boxes to resize, using something like Shift + Arrow keys would be more convenient, in my opinion. However, we also have highlight annotations that need to support both resizers.

**Comment by abaevbog** [comment]
> If it is a priority to support creating/modifying annotations, we can try to implement that first.

I don't think this is something that is super urgent. As far as VPAT goes, there are plenty of other accessibility PRs waiting to be merged and it's probably best to be thorough here. I believe the plan was to have the first VPAT report ready around the time when the school year begins, so not having this would just mean that one of the success criteria is marked as "Partially supports" instead of "Supports" with a description that annotations cannot be added/resized via keyboard at this point. Which is probably reasonable. @stakats may correct me on this and on the timeline.

> However, we need to decide exactly how that will work

So I tried to go through a number of PDF-related tools and I didn't see any unified approach to this problem. Adobe Acrobat in general seems to have the best support for keyboard navigation, and `options-arrows` can be used to resize rectangles and other figures while usual arrows move it. I didn't find a way to resize text annotations (highlight and underline) at all - via keyboard or mouse. Maybe because the idea is that you just delete and re-create an annotation instead? In any case, this would not work for our use case.


> If we only had square boxes to resize, using something like Shift + Arrow keys would be more convenient, in my opinion. However, we also have highlight annotations that need to support both resizers.

I agree that arrow keys with a modifier are easier and maybe more conventional (judging by adobe acrobat example). Would it be totally unreasonable to suggest to use `arrowLeft`/`arrowRight` to move one side of the text annotation and `arrowUp`/`arrowDown` to move the other one? While a bit unusual, it would allow us to not have to use too many modifier keys. 

Another idea I had inspired by Adobe Acrobat is to maybe add a "Resize" option into the context menu of an annotation? Then, one could activate the resizing mode @AbeJellinek described in a potentially more obvious fashion.

This is the setup I am thinking about:
- user tabs through the sidebar into the actual reader, tabs one more time to focus an annotation (with grey outline)
- uses arrows to navigate through annotations till they find one they are after
- if they want to resize/move it, they open context menu and select "Resize" (alternatively, click Enter, which selects the annotation, from where they can tab to the options button and reach the context menu from there)
- for textual annotations, `arrowUp`/`arrowDown` move the left side and `arrowRight`/`arrowLeft` move the right side of the annotation (though in this case, maybe `arrowRight`/`arrowLeft` and `option`/`alt` + `arrowRight`/`arrowLeft` is better)
- for other annotations, arrows move the annotation and `shift` + arrows (or `options`/`alt` + arrows as in Adobe) resize the bottom and right borders (where applicable)
- Enter/Escape exits resizing mode and focuses the annotation, from where arrows will change focus to the next annotation again

I feel like something like that could strike a decent balance between not adding too many additional new states while also not forcing us to add a new keyboard shortcut for every single moving/resizing action out there.

**Comment by mrtcode** [comment]
> - user tabs through the sidebar into the actual reader, tabs one more time to focus an annotation (with grey outline)
> - uses arrows to navigate through annotations till they find one they are after

In my previous comments, I mentioned wanting to eliminate the gray selection mode. However, after further consideration, I decided to make look and work more like the regular keyboard focus ring we see on buttons and inputs:
- Change its color to match other focus rings.
- Make the corners round.
- Ensure it works with all currently visible objects (annotations, links, citations). It won’t scroll the PDF view and won't focus anything beyond the currently visible view area.

This object focus mode will also be activated if you press an arrow key on a currently selected annotation, but only if the annotation was selected in the view, not in the sidebar. This is because the idea of being able to select an annotation in the view but still navigate like in a list (which also doesn’t work when the sidebar is collapsed) is weird and none of other PDF viewers do it. So I believe this approach will feel similar to when you click inside the page number input in the toolbar and then use the arrow keys to navigate to other buttons in the toolbar.

> - if they want to resize/move it, they open context menu and select "Resize" (alternatively, click Enter, which selects the annotation, from where they can tab to the options button and reach the context menu from there)

Resizing mode makes sense, but I’m not sure if it justifies adding an extra context menu option, which most people won’t need. Maybe it could be included in the Edit menu and also be triggerable with a key combination? Additionally, "Resize" is a bit misleading if you can also move the annotation, or not? Also I don’t see any "Resize" option in my version of Acrobat Reader.

> - for textual annotations, `arrowUp`/`arrowDown` move the left side and `arrowRight`/`arrowLeft` move the right side of the annotation (though in this case, maybe `arrowRight`/`arrowLeft` and `option`/`alt` + `arrowRight`/`arrowLeft` is better)

It might not be a problem for the resize mode, but take into account that <kbd>Alt-ArrowRight</kbd> and <kbd>Alt-ArrowLeft</kbd> is currenlty used on Windows and Linux for history navigation. <kbd>ArrowUp</kbd> and <kbd>ArrowDown</kbd> feels quite weird to control one of the resizers. How about using alone <kbd>ArrowLeft</kbd> and <kbd>ArrowRight</kbd> control the right side, while holding <kbd>Shift</kbd> will control the left side?

> - for other annotations, arrows move the annotation and `shift` + arrows (or `options`/`alt` + arrows as in Adobe) resize the bottom and right borders (where applicable)

Arrows for movement and <kbd>Shift</kbd> + arrows for resize sounds good. I tried this out in Acrobat Reader as well and it feels pretty good.

**Comment by abaevbog** [comment]
> In my previous comments, I mentioned wanting to eliminate the gray selection mode.

Right, yes! I was only referring to the current gray selection mode that navigates the objects on the page as opposed to selecting the annotations in the sidebar. From your last comment, I see the proper term for what I meant was "object focus mode".

 > Resizing mode makes sense, but I’m not sure if it justifies adding an extra context menu option, which most people won’t need. Maybe it could be included in the Edit menu and also be triggerable with a key combination?

That makes sense. Could be something like `Cmd-*` (where * will depend on the exact wording of the menu, I have a few more ideas below)? In Adobe Acrobat, when you place focus onto an annotation via tab, you can't immediately begin moving it with arrows. You need to sorta "select" or "activate" it with a space or Enter. This exact approach wouldn't work in our case since Enter selects the annotation but focusing an annotation with arrows and pressing a menu shortcut to start moving/resizing it seems comparable to me.

> Additionally, "Resize" is a bit misleading if you can also move the annotation, or not?

That's fair. The exact wording might need a bit or work. A few alternative ideas: "Move & Resize" (maybe a bit too wordy), "Transform" or "Arrange" (not too descriptive but I think that encompasses both "resize" and "move" aspect"), "Adjust Positioning" (I feel like "positioning" might refer to both x-y location, as well as the sizing)? Respectively, the menu shortcut could be `Cmd-R` (for "Resize"), `Cmd-M` (for "Move") and etc.

> Also I don’t see any "Resize" option in my version of Acrobat Reader.

Yeah, this suggestion was purely inspired by the "Properties" context menu option that opens a window where you can adjust some of the visual properties of the annotation. So I thought, what if instead of opening the window, it activates a mode to tweak the positioning of the object? [Forms](https://helpx.adobe.com/acrobat/using/pdf-form-field-basics.html) should have more relevant options available through that menu, such as "Positioning". It seems only available in Acrobat Pro, so I couldn't check that.

> How about using alone ArrowLeft and ArrowRight control the right side, while holding Shift will control the left side?

I think that's great! That way, we have the "resizing/moving" mode where you only use arrows or `shift`+arrows to perform operations, which I think will be easy to use and remember.

**Comment by abaevbog** [comment]
If resizing/moving annotations like this (or in a similar way) is in place, the question of adding new annotations will be almost completely solved! For area, note and text annotations, we can just drop it into a pre-defined location as discussed earlier and then move them where needed.

Highlight and underline annotations would still be a bit trickier because, as described in this [forums discussion](https://forums.zotero.org/discussion/116486/beta-7-accessibility-feedback-on-reader), in most cases, one would have the screen reader in browse mode when most keypresses are handled by screen reader and we don't know the location of the virtual cursor that reads content or have access to what it selects. So even if the [caret is added](https://github.com/zotero/zotero/issues/4224#issuecomment-2258048209) similar to Adobe Acrobat, we would still have the issue of connecting that caret with the virtual cursor of the reader. I tried a few things and wanted to share a potentially feasible workflow to handle these limitations:

1. user selects the underline or highlight annotation in the toolbar
2. user navigates the reader's content with NVDA or JAWS virtual cursor reading text
3. user hears a sentence they want to add annotation to
4. with the virtual cursor, the user selects and copies the text they are after
5. user opens "Find in Document",  pastes the copied text, maybe navigate through a few matches to find the desired result (we might need to tweak the `aria-live` announcements to make it clearer where the search result is in relation to the rest of the content)
6. finally, when the correct search result is found, they would need a way to convert a search result into an annotation of the selected type. We could automatically turn the current search result in the reader into the annotation if the underline or highlight annotation mode was selected via a keypress (as opposed to mouse)? Alternatively, there could be another shortcut involved or an explicit small "Make annotation" button in the "Find in document" popup.

**Comment by mrtcode** [comment]
So many steps. Maybe there exists a simpler way to identify text selected by a screen reader? Could copying text with a screen reader trigger the JavaScript "copy" event or something?

**Comment by mrtcode** [comment]
> We could make the new annotation keyboard shortcut add an annotation at the mouse position if the mouse is within the document.

That would only be useful for Note and Text annotations as Image annotations will need resizing anyway, right? Or could it somehow help with highlight annotations? Maybe that point could become text selection anchor and then text selection head would be controlled with arrows keys?

**Comment by abaevbog** [comment]
Sorry for the late reply!

> So many steps. Maybe there exists a simpler way to identify text selected by a screen reader? Could copying text with a screen reader trigger the JavaScript "copy" event or something?

As far as I know, everything that is done by a virtual cursor remains within the realm of the screen reader and  we do not get information or events about it. For instance, the `copy` event does not get dispatched if you select and copy text with NVDA. In a similar fashion, selecting text is not reflected in `document.getSelection()`.

I do think that I over-explained the idea a bit (e.g. step 1 isn't necessary). Here is a [recording](https://www.dropbox.com/scl/fi/61fhy3jm24vd2qam1pku2/annotation_from_search.mov?rlkey=dzznundo3wo3uucxs9zgwzepr&st=x95410v2&dl=0) with a mockup of my idea. Here, I (admittedly clunky) navigate the wikipedia article by going through a few headings, copy the last part of a paragraphs and get to "add" it as an annotation (well, to just announce it in this case). Using the search feature is an extra step but it's a fairly reliable way to bridge the gap between us and the virtual cursor.

This is not where I got this idea but Adobe Acrobat actually does do something similar. Pressing `space` or `enter` on the grippy icon to the left of the search input will create a comment for the search result, which seems somewhat similar to this workflow.

<img width="972" alt="Screenshot 2024-08-09 at 11 28 15 AM" src="https://github.com/user-attachments/assets/4537b7f7-8d93-4dc1-9d13-ea18c519b487">

----

> We could make the new annotation keyboard shortcut add an annotation at the mouse position if the mouse is within the document.

I think this would be a possible enhancement! From the [2.1.1 success criterion](https://www.w3.org/WAI/WCAG21/Understanding/keyboard#dfn-keyboard-interface) perspective, though, it might not count as a "operable through a keyboard interface" because it would rely on a mouse positioning. At the bottom of that link, under "keyboard interface" definition, it even mentions that "Operation of the application (or parts of the application) through a keyboard-operated mouse emulator, such as MouseKeys, does not qualify as operation through a keyboard interface because operation of the program is through its pointing device interface, not through its keyboard interface.". And getting a screen reader to move the mouse feels like exactly what this refers to.

Also, there can be a user relying on keyboard interface who does not use screen readers. For instance, one with a perfectly fine vision but some kind of impairment preventing them from operating a mouse. So whatever solution we go with should account for this use case as well.

**Comment by abaevbog** [comment]
> Could we write scripts/modules for NVDA and JAWS that expose the reader selection to Zotero somehow? It just makes me feel kind of bad to have to recommend a relatively lengthy workflow for the most basic operation in the reader!

I believe so!

NVDA has a full developer guide for scripts and plugins: https://www.nvaccess.org/files/nvda/documentation/developerGuide.html as well as the addon store: https://github.com/nvaccess/addon-datastore. Standalone integrations between screen readers and the reader is something I did not dive too deeply into but I believe it should be feasible with NVDA. As an open source software, there is a lot of crowd-sourced knowledge to use.

JAWS has a rich eco system of scripts and plugins as well. I wrote a couple simple ones while battling with another JAWS-related issue, so theoretically it should be doable as well. One thing with JAWS is that due to it not being open sourced, we might have to reach out to FreedomScientific for better distribution.

But the short answer is yes, it should be possible!

**Comment by mrtcode** [comment]
Yes, I agree, Find result → annotation sounds good after @abaevbog clarified it.

From a technical standpoint, I think we'll need to introduce a `createAnnotationFromFindResult(type)` function in both PDF and DOM views, which can later be called from the find popup.

But the question is, how do you choose which annotation to create while being in the find popup?

**Comment by abaevbog** [comment]
> But the question is, how do you choose which annotation to create while being in the find popup?

Because this only applies to two types of annotations (highlight and underline), and you can easily convert one into another via an option in the context menu, I think we have some flexibility here. What if the user doesn't have to choose anything and we just go with the most-recently-used annotation type? That way, if it's the right type - great. If not, you can change it via context menu, and next time, it will be right. Maybe we also add an extra check to see if an "annotation" or "highlight" tool is selected, in which case we should go with the selected type regardless of what last added annotation was.

I feel like this would allow us to strike a reasonable balance between not requiring an additional step of selecting annotation type for keyboard users, while also allowing to easily override this behavior if needed?

**Comment by mrtcode** [comment]
So we still need a keyboard shortcut. What it could be? Maybe we should have two different ones for different annotation types, or would that be too complicated?

**Comment by abaevbog** [comment]
My initial thought was we could make it an actual button that one can tab onto and click from the popup, as opposed to it just being a shortcut. The main reason being that shortcuts are not easily discoverable and you have to remember them. Also, once there is a shortcut, it's quite hard to change it because then everyone thinks the shortcut stopped working and reports it as a bug (e.g. as in [collection highlighting on linux](https://github.com/zotero/zotero/issues/4386)). So I thought we could do something like this:

<img width="334" alt="Screenshot 2024-08-14 at 10 29 17 AM" src="https://github.com/user-attachments/assets/fb60a401-cc09-4979-85ad-0c836b1a7d60">

Now, though, I am thinking that if we were to go this way, why not just add two buttons, each for the annotation type. Then, the whole logic of keeping track of the last used annotation is just redundant and the interface is more explicit. So I think you're right that just giving 2 separate options to pick which annotation you want is better.

<img width="340" alt="Screenshot 2024-08-14 at 10 42 06 AM" src="https://github.com/user-attachments/assets/afdaabfa-861b-492c-9b34-1b0711342c7b">

If we do want a shortcut (either instead of buttons or as a complement), how about `Alt-1/2` for windows and `Ctrl-1/2` for mac? `Alt-1/2` corresponds to selecting the annotation tool in the actual reader, so it seems somewhat reasonable if it were to create an annotation of the respective type when the popup is opened. On mac, following the same logic, we would want to do `Option-1/2` but it actually prints characters "¡" and "™" in the input field, and I'm not sure we should be interfering with that. So `Ctrl-1/2` feels like the next best thing? 

Given that these shortcuts are fairly intuitive (meaning, one might even guess that), I think discoverability is less of an issue here than I thought and we're unlikely to ever want to change these shortcuts specifically when the search popup is active. So maybe the usual drawbacks of shortcuts mentioned above don't apply here as much, and it's just a matter of what we think would look/feel nicer. But whichever way we go, I now feel like keeping track of the last-used annotation type as suggested earlier is not worth it and it is better to add either a separate button or a separate shortcut for each annotation type (or both!).

**Comment by mrtcode** [comment]
I think adding extra buttons to the find popup might be too prominent since most people won’t need them.

Using two keyboard shortcuts seems like a better approach. We could make this more versatile by allowing users to create highlights or underlines not just for the current search result, but also for manually selected text.

Additionally, we could introduce shortcuts for Note, Text, and Image annotations, allowing users to insert them, for example, in the middle of the current page as we discussed before.

As we use Alt/Option + 1, 2, 3, etc., as shortcuts for activating toolbar buttons, it would make sense to use the same shortcut with an additional modifier to trigger the button action. But the only modifier that doesn’t seem to interfere with text input on macOS is Ctrl (not sure about Windows/Linux). So, how about Ctrl + Alt/Option + 1, 2, 3, etc.?

**Comment by abaevbog** [comment]
> I think adding extra buttons to the find popup might be too prominent since most people won’t need them.

That's fair.

> Using two keyboard shortcuts seems like a better approach. We could make this more versatile by allowing users to create highlights or underlines not just for the current search result, but also for manually selected text.

True! Also if the cursor is added similar to adobe acrobat, it will be a nice addition to select text and turn it into annotation with that shortcut. 

> So, how about Ctrl + Alt/Option + 1, 2, 3, etc.?

That seems reasonable! I checked the documentation on Voiceover, NVDA and JAWS shortcuts, and there does not seem to be any conflict with these.

**Comment by abaevbog** [comment]
Let me sum up the keyboard interface we have arrived at so far:
- to add a highlight or underline annotation to a piece of text in the document, user opens the "Find in document" popup, types the text to find it, and uses `Ctrl + Alt/Option + 1, 2` to highlight/underline text for the selected search result as discussed [here](https://github.com/zotero/zotero/issues/4224#issuecomment-2286266706)
- to add a note, text, image annotation, user can use `Ctrl + Alt/Option + 3, 4, 5` to drop the note, text, area onto the current page as discussed [here](https://github.com/zotero/zotero/issues/4224#issuecomment-2291737618)
- to edit an annotation, user selects the annotation with arrows in object focus mode, selects Edit > Resize (named better) menu to turn on "Resizing" mode as discussed [here](https://github.com/zotero/zotero/issues/4224#issuecomment-2273416134). Options for the menu name that I had: "Move & Resize", "Transform" or "Arrange", "Adjust Positioning". Then:
    - for highlight/underline annotations, ArrowLeft and ArrowRight move the right side, and Shift-ArrowLeft/Right control the left side
    - for note, text and image annotation, arrows will move the annotation. For image annotation, Shift-Arrows will also move the borders resizing them
 
Correct me if anything doesn't sound right in this summary

**Comment by mrtcode** [comment]
That's a great sum up.

A few more things:
- Ink annotation will also have the same controls as image annotation, except the width/height ratio will be fixed.
- Text annotation will also support Shift-ArrowLeft/ArrowRight for resizing the right side.
- Note, text, and image annotations will probably be added at the center of the currently visible PDF view (even if it's zoomed in). For the DOM view, it should add a note annotation to the closest slot near the view center.
- Should we have a keyboard shortcut to enter into the resizing mode?

**Comment by abaevbog** [comment]
> Should we have a keyboard shortcut to enter into the resizing mode?

I suppose it wouldn't hurt, the natural question is what would a good candidate be? Something like `Cmd/Alt-R` (for "resize")?

And it is not strictly required, since you very much can navigate the menu with a keyboard but it likely would save some time.

**Comment by mrtcode** [comment]
> to edit an annotation, user selects the annotation with arrows in object focus mode, selects Edit > Resize (named better) menu to turn on "Resizing" mode as discussed [here](https://github.com/zotero/zotero/issues/4224#issuecomment-2273416134). Options for the menu name that I had: "Move & Resize", "Transform" or "Arrange", "Adjust Positioning". Then:
> * for highlight/underline annotations, ArrowLeft and ArrowRight move the right side, and Shift-ArrowLeft/Right control the left side
> * for note, text and image annotation, arrows will move the annotation. For image annotation, Shift-Arrows will also move the borders resizing them

But once again, why do we need the resizing mode at all? Why not just to have i.e. `Shift-ArrowKeys` and `Shift-Option/Alt-ArrowKeys` and achieve exactly the same move/resize operations but without the extra mode which looks like an unnecessary complexity?

One of the arguments was that it might be hard to remember the keyboard shortcuts, but the resizing mode doesn’t reduce the number of shortcuts—it just simplifies them slightly (e.g., ArrowKeys vs. Shift-Option/Alt-ArrowKeys). However, enabling the resizing mode adds an extra step, so I don’t see much benefit in it.

**Comment by abaevbog** [comment]
I suppose you're right... I think the perceived benefit of the resizing mode was to avoid the complexity of using a large number of shortcuts and to have a unified interface for these operations. But I see that we are arriving at a number of shortcuts that are all needed with or without the resizing mode, so it does seem like an extra step at this point. If we were to drop it, is that how the moving/resizing could work?

- to move a text, note, image or ink annotation across the reader, you could use `Shift-Arrows` (this is consistent with Adobe Acrobat where holding `shift` while pressing arrow keys move the object faster than without `shift`). 
- to resize an image or text annotation, you would use `Shift-Option/Alt-ArrowLeft/Right` for the right edge. For image annotation,  you can also use `Shift-Option/Alt-ArrowUp/Down` for the bottom. 
- for highlight and underline annotations, `Shift-ArrowRight/Left` could move the right edge and `Shift-Option/Alt-ArrowLeft/Right` could move the left edge of the annotation.

That does sound way easier to me!

**Comment by abaevbog** [comment]
Yeah, I fully agree with the modal system approach as a general rule! With the keyboard accessibility being a part of the focus here, I know that "let's just add a shortcut for that" is often an easy way out, and if done without proper consideration this can lead to a confusing interface and an unnecessary number of shortcuts floating around. So I always think it's always worth exploring non-shortcut approach if possible.

But here it does seem to come together as one fairly cohesive interface, and it seems like the most convenient way for keyboard users to make selections and to navigate.

In addition, I feel like dropping the resizing mode might make shortcuts more useful for more people. People who are not forced to use a keyboard are almost guaranteed to never use the resizing mode because of the extra step involved, however they might use the shortcut now. Like if you want to extend the annotation by a single character (I often miss punctuation signs during highlight and then edit the annotation to include it), it might be faster to click `Shift-Option/Alt-ArrowRight` than to drag.


The only thing is that now when an annotation is selected in the sidebar, `Shift-ArrowUp/Down` is responsible for selecting multiple annotations. This does not apply to the object focus mode (when a grey outline appears right now), so I don't think there are any issues here. But still worth mentioning, just to be sure.

**Comment by mrtcode** [comment]
Yes, the keyboard shortcuts list looks good. I'm implementing it now, and I'll push a PR later so we can try it out. Once everything looks good, we can implement the same for the DOM view.

Just one thing. Shift-Option-ArrowLeft/ArrowRight is used for word selection on macOS, therefore we might want to consider other alternatives for this keyboard shortcut.

**Comment by abaevbog** [comment]
Awesome!

> Just one thing. Shift-Option-ArrowLeft/ArrowRight is used for word selection on macOS, therefore we might want to consider other alternatives for this keyboard shortcut.

I don't know what the convention is in this regard. Our shortcuts would only apply when an annotation is focused, so there should never be a conflict. So would it be an issue? If we do want an alternative, maybe `Shift+Command+Arrow`? I don't think this is being used anywhere at this point.
